src/EventManagement.Domain/Registration.cs

```diff
@@ -156,60 +156,96 @@ public void CreateOrder(IEnumerable<Product> products, IEnumerable<ProductVarian
 		/// Updates an existing order if it's not already been invoiced.
 		/// Else creates a new order.
 		/// </summary>
 		/// <param name="products"></param>
 		/// <param name="variants"></param>
-		public void CreateOrUpdateOrder(IEnumerable<Product> products, IEnumerable<ProductVariant> variants)
+		public void CreateOrUpdateOrder(ICollection<Product> products, ICollection<ProductVariant> variants)
 		{
-			// Check if the product already exists in one of this registration's orders
-			var existingProductIds = Orders.Where(o => o.Status != OrderStatus.Cancelled)
-											.SelectMany(o => o.OrderLines
-											.Where(l => l.ProductId.HasValue)
-											.Select(l => l.ProductId.Value)
-										);
-			if(existingProductIds.Intersect(products.Select(p => p.ProductId)).Any())
-			{
-				throw new InvalidOperationException("The same product cannot be added twice");
-			}
+			// Get the existing productids
+            var existingProducts = Orders.Where(o => o.Status != OrderStatus.Cancelled && o.Status != OrderStatus.Refunded)
+                                        .SelectMany(o => o.OrderLines
+                                        .Where(l => l.ProductId.HasValue)
+                                    );
+			var existingProductIds = existingProducts.Select(l => l.ProductId.Value);
+
+            // Check if an order needs to be refunded
+            var conflictingProductIds = existingProductIds.Intersect(products.Select(p => p.ProductId));
+            var conflictingProducts = existingProducts
+				.Where(p => conflictingProductIds.Contains(p.ProductId.Value));
+
+			// If a refund is required
+			var refundLines = new List<OrderLine>();
+            if(conflictingProductIds.Any())
+            {
+				var ordersToRefund = Orders
+								.Where(o => o.OrderLines.Where(l => conflictingProductIds.Contains(l.ProductId.Value)).Any())
+								.GroupBy(o => o.OrderId)
+								.Select(g => g.First());
+				
+				// Refund the orders, and create refundlines for each of them
+				foreach(var o in ordersToRefund)
+				{
+					o.MarkAsRefunded();
+					refundLines.Add(OrderLine.CreateRefundOrderLine(o));
+				}
+
+				// Update the produces, variants arguments
+				var refundedOrdersLines = ordersToRefund.SelectMany(l => l.OrderLines);
+				foreach(var line in refundedOrdersLines)
+				{
+					if(!products.Where(p => p.ProductId == line.ProductId).Any())
+					{
+						products.Add(line.Product);
+						if(line.ProductVariantId != null)
+						{
+							variants.Add(line.ProductVariant);
+						}
+					}
+				}
+            }
 
 			// Check if any uninvoiced orders exist (excluding the cancelled ofcourse)
-			var uninvoicedOrders = Orders.Where(o => o.Status != OrderStatus.Invoiced && o.Status != OrderStatus.Cancelled);
+			var editableOrders = Orders.Where(o => o.Status != OrderStatus.Invoiced && o.Status != OrderStatus.Cancelled && o.Status != OrderStatus.Refunded);
 
-			if(!uninvoicedOrders.Any())
+			if(!editableOrders.Any())
```

Comment:
No editable orders exist?

Reply:
Cancelled, Invoiced & Refunded orders cannot be edited. That condition is to decide if I should make the changes to an existing order or if I should create a new order.

The order is editable if it's `Draft` or `Verified`:

https://github.com/gldraphael/EventManagement/blob/ef91550ec1d6293e9fcaddb465f88977828596f7/src/EventManagement.Domain/Order.cs#L104-L105